cancelable/closer.go

```diff
@@ -12,24 +12,65 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// Package cancelable provides a cancelable version of io.Closer.
+//
+// This is intended to be used by functions where the responsibility for closing
+// a resource sometimes belongs to the function itself (e.g. early returns due
+// to error handling), and sometimes belongs to the caller.
+//
+// In the example below, this responsibility is shared by both f() and caller().
+// There is always a deferred function that is ready to close the connection
+// when needed, and we never need to worry about forgetting to call Close().
+//
+//  func f() (net.Conn, error) {
+//      conn, err := net.Dial("example.com:443")
+//      if err != nil {
+//          return nil, err
+//      }
+//      // If something goes wrong, f() is responsible for closing conn.
+//      closer := cancelable.Closer(conn)
+//      defer closer.Close()
+//      if err := frob(conn); err != nil {
+//          // Early exit; conn will be closed by the cancelable.Closer.
+//          return nil, err
+//      }
+//      // After the return, it is the caller's responsibility to close conn.
+//      closer.Cancel()
+//      return conn, nil
+//      // The defered closer.Close() still happens, but doesn't do anything.
+//  }
+//
+//  func caller() {
+//      conn, err := f()
+//      if err != nil {
+//          return nil, err
+//      }
+//      defer conn.Close()
+//      // ...
+//  }
 package cancelable
 
 import (
 	"io"
 )
 
+// Closer is an io.Closer that can be cancelled.
 type Closer struct {
 	c io.Closer
 }
 
+// NewCloser returns a cancelable.Closer that wraps an io.Closer.
 func NewCloser(c io.Closer) *Closer {
 	return &Closer{c: c}
 }
 
+// Cancel turns any future calls to Close into a no-op.
 func (c *Closer) Cancel() {
 	c.c = nil
 }
 
+// Close calls the Close function on the io.Closer. It is safe to call this
```

Comment:
Not necessarily, only if it has not been cancelled or already closed.

Reply:
Done